File: java/robot-simulator/src/main/java/GridPosition.java
class GridPosition {
    final int x;
    final int y;

    GridPosition(int x, int y) {
        this.x = x;
        this.y = y;
    }

    GridPosition move(Orientation orientation) {
        if (orientation == Orientation.NORTH) {
            return new GridPosition(x, y + 1);
        } else if (orientation == Orientation.EAST) {
            return new GridPosition(x + 1, y);
        } else if (orientation == Orientation.SOUTH) {
            return new GridPosition(x, y - 1);
        } else {
            return new GridPosition(x - 1, y);
        }
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + x;
        result = prime * result + y;
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        } else if (obj == null || getClass() != obj.getClass()) {
            return false;
        } else if (x != ((GridPosition) obj).x || y != ((GridPosition)obj).y) {
            return false;
        } else {
            return true;
        }
    }
}

File: kotlin/robot-simulator/src/main/kotlin/GridPosition.kt
data class GridPosition(val x: Int, val y: Int) {
    fun move(orientation: Orientation): GridPosition = when (orientation) {
        Orientation.NORTH -> GridPosition(this.x, this.y + 1)
        Orientation.SOUTH -> GridPosition(this.x, this.y - 1)
        Orientation.EAST -> GridPosition(this.x + 1, this.y)
        Orientation.WEST -> GridPosition(this.x - 1, this.y)
    }
}
